pkg/yang/entry.go

```diff
@@ -675,34 +727,120 @@ func ToEntry(n Node) (e *Entry) {
 				e.merge(nil, nil, ToEntry(a))
 			}
 		case "type":
-			// We don't expect this to happen, so throw an error.
-			// BUG(borman): I think a deviate statement might trigger this.
-			e.addError(fmt.Errorf("%s: unexpected type in %s:%s", Source(n), n.Kind(), n.NName()))
+			// The type keyword is specific to deviate to change a type. Other type handling
+			// (e.g., leaf type resolution) is done outside of this case.
+			n, ok := n.(*Deviate)
+			if !ok {
+				e.addError(fmt.Errorf("unexpected type found, only valid under Deviate, is %T", n))
+				continue
+			}
 
+			if n.Type != nil {
+				if errs := n.Type.resolve(); errs != nil {
+					e.addError(fmt.Errorf("deviation has unresolvable type, %v", errs))
+					continue
+				}
+				e.Type = n.Type.YangType
+			}
+			continue
 		// Keywords that do not need to be handled as an Entry as they are added
 		// to other dictionaries.
-		case "default",
-			"typedef":
+		case "default":
+			if e.Kind == LeafEntry {
+				// default is handled separately for a leaf, but in a deviate statement
+				// we must deal with it here.
+				continue
+			}
+			d, ok := fv.Interface().(*Value)
+			if !ok {
+				e.addError(fmt.Errorf("%s: unexpected default type in %s:%s", Source(n), n.Kind(), n.NName()))
+			}
+			e.Default = d.asString()
+		case "typedef":
 			continue
 		case "deviation":
 			if a := fv.Interface().([]*Deviation); a != nil {
 				for _, d := range a {
+					e.Deviations = append(e.Deviations, &DeviatedEntry{
+						Entry:        ToEntry(d),
+						DeviatedPath: d.Statement().Argument,
+					})
+
 					for _, sd := range d.Deviate {
 						if sd.Type != nil {
 							sd.Type.resolve()
 						}
 					}
 				}
 			}
+		case "deviate":
+			if a := fv.Interface().([]*Deviate); a != nil {
+				for _, d := range a {
+					de := ToEntry(d)
+					var arg deviationType
+					switch d.Statement().Argument {
+					case "not-supported":
+						arg = DeviationNotSupported
+					case "add":
+						arg = DeviationAdd
+					case "replace":
+						arg = DeviationReplace
+					case "delete":
+						arg = DeviationDelete
+					}
+
+					if e.Deviate == nil {
+						e.Deviate = map[deviationType][]*Entry{}
+					}
+
+					if _, ok := e.Deviate[arg]; !ok {
```

Comment:
Since the map is of value types, I believe you can avoid the existence check and slice creation (796-798) entirely, because the `[]*Entry{}(nil)`  you get back from the `e.Deviate[arg]` map lookup on line 799 is valid to be appended to.

Reply:
Thanks! Fixed.